### PR TITLE
Add Basis swap helper for yield curve bootstrapping

### PR DESCRIPTION
**Pull request by jiskin**
I didn't find any equivalent on Quantlib project, so I created this one.

### CODE REVIEW

**Comment by igitur** [comment]
`BasisSwap` also doesn't exist in QuantLib. @amaggiulli , mind if I port it back?

**Comment by igitur** [comment]
Any unit tests that we can add for this?

**Comment by jiskin** [comment]
Hello igitur,

There is already an equivalent from QuantExt which is based on QuantLib: https://github.com/OpenSourceRisk/Engine/blob/master/QuantExt/qle/instruments/tenorbasisswap.cpp

For the unit test I should add a bootstrapping with Bloomberg comparison, but I currently only have time to create the instruments.

File: QLNet/Instruments/BasisSwap.cs [hunk 1]
```diff
@@ -219,10 +243,41 @@ protected override void setupExpired()
...
+            if (legBPS_[longNo_] != null) {
+                fairLongSpread_ = longSpread_ - NPV_ / (legBPS_[longNo_] / basisPoint);
+            }
+        }
+
+        /* Short fair spread calculation ok if no averaging/compounding OR
+           if there is averaging/compounding and the spread is added after */
+        if (fairShortSpread_ == null) 
+        {
+                if (legBPS_[shortNo_] != null)
```
Comment (amaggiulli) [inline QLNet/Instruments/BasisSwap.cs]: MAJOR Merge this if statement with the enclosing one. [rule](https://sonarqube.com/coding_rules#rule_key=csharpsquid%3AS1066)

File: QLNet/Instruments/BasisSwap.cs [hunk 2]
```diff
@@ -219,10 +243,41 @@ protected override void setupExpired()
...
+         }
+         else
+         {
+             fairLongSpread_ = null;
+             fairShortSpread_ = null;
+         }
+
+        // Long fair spread should be fine - no averaging or compounding
+        if (fairLongSpread_ == null) {
+            if (legBPS_[longNo_] != null) {
```
Comment (amaggiulli) [inline QLNet/Instruments/BasisSwap.cs]: MAJOR Merge this if statement with the enclosing one. [rule](https://sonarqube.com/coding_rules#rule_key=csharpsquid%3AS1066)

File: QLNet/Instruments/BasisSwap.cs [hunk 3]
```diff
@@ -42,6 +42,9 @@ public enum Type { Receiver = -1, Payer = 1 };
       private DayCounter floating2DayCount_;
       private BusinessDayConvention paymentConvention_;
 
+      private int longNo_, shortNo_;
```
Comment (amaggiulli) [inline QLNet/Instruments/BasisSwap.cs]: MAJOR Make "longNo_" "readonly". [rule](https://sonarqube.com/coding_rules#rule_key=csharpsquid%3AS2933)
MAJOR Make "shortNo_" "readonly". [rule](https://sonarqube.com/coding_rules#rule_key=csharpsquid%3AS2933)

File: QLNet/Instruments/BasisSwap.cs [hunk 4]
```diff
@@ -42,6 +42,9 @@ public enum Type { Receiver = -1, Payer = 1 };
...
+      private double longSpread_, shortSpread_;
```
Comment (amaggiulli) [inline QLNet/Instruments/BasisSwap.cs]: MAJOR Make "longSpread_" "readonly". [rule](https://sonarqube.com/coding_rules#rule_key=csharpsquid%3AS2933)
MAJOR Make "shortSpread_" "readonly". [rule](https://sonarqube.com/coding_rules#rule_key=csharpsquid%3AS2933)

**Comment by amaggiulli** [comment]
SonarQube analysis reported 6 issues
* ![MAJOR][MAJOR] 6 major

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

File: QLNet/Instruments/BasisSwap.cs (hunk 1 above)
Comment (amaggiulli) [inline QLNet/Instruments/BasisSwap.cs]: code must be : 

```
if (fairShortSpread_ == null && legBPS_[shortNo_] != null ) 
{
...
)
```

File: QLNet/Instruments/BasisSwap.cs (hunk 2 above)
Comment (amaggiulli) [inline QLNet/Instruments/BasisSwap.cs]: code must be : 

```
if (fairLongSpread_ == null && legBPS_[longNo_] != null) 
{
 ...
}
```

**Pull request closed without merging**